File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/AutonomousMode.java
/* Copyright (c) 2017 FIRST. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted (subject to the limitations in the disclaimer below) provided that
 * the following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this list
 * of conditions and the following disclaimer.
 *
 * Redistributions in binary form must reproduce the above copyright notice, this
 * list of conditions and the following disclaimer in the documentation and/or
 * other materials provided with the distribution.
 *
 * Neither the name of FIRST nor the names of its contributors may be used to endorse or
 * promote products derived from this software without specific prior written permission.
 *
 * NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE GRANTED BY THIS
 * LICENSE. THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package org.firstinspires.ftc.teamcode;

import com.qualcomm.robotcore.eventloop.opmode.Autonomous;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.OpMode;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.util.ElapsedTime;

import org.firstinspires.ftc.robotcore.external.ClassFactory;
import org.firstinspires.ftc.robotcore.external.navigation.RelicRecoveryVuMark;
import org.firstinspires.ftc.robotcore.external.navigation.VuforiaLocalizer;
import org.firstinspires.ftc.robotcore.external.navigation.VuforiaTrackable;
import org.firstinspires.ftc.robotcore.external.navigation.VuforiaTrackables;

import java.util.Locale;
import java.util.concurrent.FutureTask;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

@Autonomous(name = "Autonomous program")
public class AutonomousMode extends LinearOpMode {
    private ElapsedTime runtime = new ElapsedTime();
    private volatile String logStr = "";
    private synchronized void log(String text) {
        logStr += "[" + runtime.time(TimeUnit.SECONDS) + "] " + text + "\n";
        telemetry.addLine(logStr);
        telemetry.update();
    }

    private void sleepSync() throws OpModeStoppedException {
        if (!opModeIsActive())
            throw new OpModeStoppedException();
        else
            idle();
    }

    private void turnSync(Hardware hw, int degrees) throws OpModeStoppedException {
        int encoderTicks = (int)Math.round(degrees * Hardware.TETRIX_TICKS_PER_TURN_DEGREE);
        DcMotor.RunMode oldMode = hw.leftDriveMotor.getMode();

        hw.leftDriveMotor.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
        hw.leftDriveMotor.setMode(DcMotor.RunMode.RUN_TO_POSITION);

        hw.leftDriveMotor.setTargetPosition(encoderTicks);
        log(String.format(Locale.US,
                "Turning %d degrees, which is %d ticks", degrees, encoderTicks));
        hw.leftDriveMotor.setPower(degrees >= 0 ? 0.1 : -0.1);
        hw.rightDriveMotor.setPower(degrees >= 0 ? -0.1 : -0.1);
        while (hw.leftDriveMotor.isBusy()) {
            sleepSync();
        }
        hw.leftDriveMotor.setPower(0);
        hw.rightDriveMotor.setPower(0);
        hw.leftDriveMotor.setMode(oldMode);
    }

    private RelicRecoveryVuMark detectPictogram() throws OpModeStoppedException {
        int cameraMonitorViewId = hardwareMap.appContext.getResources().getIdentifier("cameraMonitorViewId", "id", hardwareMap.appContext.getPackageName());
        VuforiaLocalizer.Parameters parameters = new VuforiaLocalizer.Parameters(cameraMonitorViewId);
        parameters.vuforiaLicenseKey = "AcdD/rP/////AAAAGQcYKmwTDk0lulf4t6n2JsQiodu68wCwukVguR/SeZyNkVD0OnUmmSWSrpM2jXTVVNorEhJRyV08URkTRak94XQN8/jPzVxzuOLCQ8VR8uYKuP/JoovnJM2MC3Pc1KvLlrLwWrL4185vpVaQMLRmvCkzNH+lyoEusMC7vwT4ayI6I22ceFumQuAubLp8APiT3omF4KG6W/lqNyJukt9YHgBYO/JJRVPfZg04LEhwFMixYOXfh+moWdf8zCMj+V7GUfH7Q7OGM0jobzVrg0uYboA2nrJBRjQS6j2eGoXX4yRwhmeVLVtBuklgw+n3qXgQ+OX9Lp48xNIApOByAlAhU117gDYYwE5NQ8ADKvtgupKd";
        parameters.cameraDirection = VuforiaLocalizer.CameraDirection.BACK;
        VuforiaLocalizer vuforia = ClassFactory.createVuforiaLocalizer(parameters);
        VuforiaTrackables relicTrackables = vuforia.loadTrackablesFromAsset("RelicVuMark");
        VuforiaTrackable relicTemplate = relicTrackables.get(0);
        relicTemplate.setName("relicVuMarkTemplate"); // can help in debugging; otherwise not necessary
        relicTrackables.activate();
        while (true) {
            RelicRecoveryVuMark vuMark = RelicRecoveryVuMark.from(relicTemplate);
            if (vuMark != RelicRecoveryVuMark.UNKNOWN) {
                return vuMark;
            } else {
                sleepSync();
            }
        }
    }

    @Override
    public void runOpMode() {
        Hardware hw = new Hardware(hardwareMap);

        hw.leftGrabberServo.setPosition(Hardware.GRABBER_GRABBED);
        hw.rightGrabberServo.setPosition(Hardware.GRABBER_GRABBED);

        // wait for the start button to be pressed.
        waitForStart();

        runtime.reset();

        // 1. Drive straight into wall
        // 2. Scan pictogram
        // 3. Hit jewel
        // 4. Put block we already have in cryptobox
        // ...

        try {
            // TODO: Add moving back and forth if neither detected
            FutureTask<Void> jewelTask = new FutureTask<>(() -> {
                hw.jewelArmServo.setPosition(Hardware.JEWEL_ARM_EXTENDED);
                log("Starting jewel task");
                sleep(2000);

                int direction = hw.jewelColorSensor.blue() > 0
                        ? -1  // left
                        :  1; // right
                turnSync(hw, 10 * direction);
                turnSync(hw, -10 * direction);
                hw.jewelArmServo.setPosition(Hardware.JEWEL_ARM_HALF_EXTENDED);
                sleep(1000);
                return null;
            });
            // Max 10 sec
            FutureTask<RelicRecoveryVuMark> detectGlyphTask =
                    new FutureTask<>(this::detectPictogram);
            hw.lifterMotor.setTargetPosition(Hardware.LIFTER_TOP_LIMIT / 2);
            hw.lifterMotor.setPower(-0.1);
            jewelTask.run();
            detectGlyphTask.run();

            // wait for jewel task to finish
            jewelTask.get();
            log("Jewel task finished");
            // Get (blocking) glyph column
            RelicRecoveryVuMark correctGlyphColumn;
            try {
                correctGlyphColumn = detectGlyphTask.get(10, TimeUnit.SECONDS);
            } catch (TimeoutException e) {
                log("Failed to find glyph");
                correctGlyphColumn = RelicRecoveryVuMark.UNKNOWN;
            }
            log("Got glyph column " + correctGlyphColumn);

            // reeeveerse
            hw.leftDriveMotor.setPower(-0.1);
            hw.rightDriveMotor.setPower(-0.1);
            // Detect with ODS
            int columnsToPass;
            switch (correctGlyphColumn) {
                case LEFT:
                    columnsToPass = 3;
                    break;
                case RIGHT:
                    columnsToPass = 2;
                    break;
                case CENTER:
                    columnsToPass = 1;
                    break;
                case UNKNOWN:
                default:
                    columnsToPass = 2; // lmao
            }
            //hw.horizontalDriveMotor.setPower(0.25);
            int columnsPassed = 0;
            while (columnsToPass > columnsPassed) {
                while (hw.ods.getLightDetected() == 0) {
                    sleepSync();
                }
                // we have hit a glyph column wall
                columnsPassed += 1;
            }

            // Now we are at the required column. Turn & move forward until ODS reads
            hw.jewelArmServo.setPosition(Hardware.JEWEL_ARM_RETRACTED);
            turnSync(hw, -90);
            hw.leftDriveMotor.setPower(0.1);
            hw.rightDriveMotor.setPower(0.1);
            sleep(500);
            hw.rightDriveMotor.setPower(0);
            hw.leftDriveMotor.setPower(0);
            // Hopefully we've hit the box by now. Release the box!
            hw.rightGrabberServo.setPosition(Hardware.GRABBER_RELEASED);
            hw.leftGrabberServo.setPosition(Hardware.GRABBER_RELEASED);
        } catch (Exception e) {
            log("Stopping op mode... " + e);
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/AutonomousMode.kt
/* Copyright (c) 2017 FIRST. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted (subject to the limitations in the disclaimer below) provided that
 * the following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this list
 * of conditions and the following disclaimer.
 *
 * Redistributions in binary form must reproduce the above copyright notice, this
 * list of conditions and the following disclaimer in the documentation and/or
 * other materials provided with the distribution.
 *
 * Neither the name of FIRST nor the names of its contributors may be used to endorse or
 * promote products derived from this software without specific prior written permission.
 *
 * NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE GRANTED BY THIS
 * LICENSE. THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package org.firstinspires.ftc.teamcode

import com.disnodeteam.dogecv.CameraViewDisplay
import com.disnodeteam.dogecv.DogeCV
import com.disnodeteam.dogecv.Dogeforia
import com.disnodeteam.dogecv.detectors.roverrukus.GoldAlignDetector
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import com.qualcomm.robotcore.eventloop.opmode.Autonomous
import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.util.ElapsedTime
import org.firstinspires.ftc.robotcore.external.matrices.OpenGLMatrix
import org.firstinspires.ftc.robotcore.external.navigation.*
import java.util.Collections.addAll


/**
 * This file contains an minimal example of a Linear "OpMode". An OpMode is a 'program' that runs in either
 * the autonomous or the teleop period of an FTC match. The names of OpModes appear on the menu
 * of the FTC Driver Station. When an selection is made from the menu, the corresponding OpMode
 * class is instantiated on the Robot Controller and executed.
 *
 * This particular OpMode just executes a basic Tank Drive Teleop for a two wheeled robot
 * It includes all the skeletal structure that all linear OpModes contain.
 *
 * Use Android Studios to Copy this Class, and Paste it into your team's code folder with a new name.
 * Remove or comment out the @Disabled line to add this opmode to the Driver Station OpMode list
 */
@Autonomous(name = "Autonomous mode")
class AutonomousMode : LinearOpMode() {
    private val runtime = ElapsedTime()

    fun initializeVuforia(): List<VuforiaTrackable> {
        val cameraMonitorViewId = hardwareMap.appContext.resources.getIdentifier("cameraMonitorViewId", "id", hardwareMap.appContext.packageName)
        val params = VuforiaLocalizer.Parameters()
        params.apply {
            vuforiaLicenseKey = "AWbfTmn/////AAABmY0xuIe3C0RHvL3XuzRxyEmOT2OekXBSbqN2jot1si3OGBObwWadfitJR/D6Vk8VEBiW0HG2Q8UAEd0//OliF9aWCRmyDJ1mMqKCJZxpZemfT5ELFuWnJIZWUkKyjQfDNe2RIaAh0ermSxF4Bq77IDFirgggdYJoRIyi2Ys7Gl9lD/tSonV8OnldIN/Ove4/MtEBJTKHqjUEjC5U2khV+26AqkeqbxhFTNiIMl0LcmSSfugGhmWFGFtuPtp/+flPBRGoBO+tSl9P2sV4mSUBE/WrpHqB0Jd/tAmeNvbtgQXtZEGYc/9NszwRLVNl9k13vrBcgsiNxs2UY5xAvA4Wb6LN7Yu+tChwc+qBiVKAQe09\n"
            fillCameraMonitorViewParent = true
        }
        val vuforia = Dogeforia(params)
        var allTrackables: List<VuforiaTrackable>? = null
        vuforia.apply {
            enableConvertFrameToBitmap()
            val targetsRoverRuckus = loadTrackablesFromAsset("RoverRuckus")
            val blueRover = targetsRoverRuckus[0]
            blueRover.name = "Blue-Rover"
            val redFootprint = targetsRoverRuckus[1]
            redFootprint.name = "Red-Footprint"
            val frontCraters = targetsRoverRuckus[2]
            frontCraters.name = "Front-Craters"
            val backSpace = targetsRoverRuckus[3]
            backSpace.name = "Back-Space"

            val mmPerInch = 25.4f
            val mmFTCFieldWidth = 12 * 6 * mmPerInch
            val mmTargetHeight = 6 * mmPerInch

            // For convenience, gather together all the trackable objects in one easily-iterable collection */

            allTrackables = targetsRoverRuckus.toList()

            val blueRoverLocationOnField = OpenGLMatrix
                    .translation(0f, mmFTCFieldWidth, mmTargetHeight)
                    .multiplied(Orientation.getRotationMatrix(
                            AxesReference.EXTRINSIC, AxesOrder.XYZ, AngleUnit.DEGREES,
                            90.0f, 0.0f, 0.0f))
            blueRover.location = blueRoverLocationOnField

            val redFootprintLocationOnField = OpenGLMatrix
                    .translation(0f, -mmFTCFieldWidth, mmTargetHeight)
                    .multiplied(Orientation.getRotationMatrix(
                            AxesReference.EXTRINSIC, AxesOrder.XYZ, AngleUnit.DEGREES,
                            90.0f, 0.0f, 180.0f))
            redFootprint.location = redFootprintLocationOnField

            val frontCratersLocationOnField = OpenGLMatrix
                    .translation(-mmFTCFieldWidth, 0f, mmTargetHeight)
                    .multiplied(Orientation.getRotationMatrix(
                            AxesReference.EXTRINSIC, AxesOrder.XYZ, AngleUnit.DEGREES,
                            90.0f, 0.0f, 90.0f))
            frontCraters.location = frontCratersLocationOnField

            val backSpaceLocationOnField = OpenGLMatrix
                    .translation(mmFTCFieldWidth, 0f, mmTargetHeight)
                    .multiplied(Orientation.getRotationMatrix(
                            AxesReference.EXTRINSIC, AxesOrder.XYZ, AngleUnit.DEGREES,
                            90.0f, 0.0f, -90.0f))
            backSpace.location = backSpaceLocationOnField

            val CAMERA_FORWARD_DISPLACEMENT = 110   // eg: Camera is 110 mm in front of robot center
            val CAMERA_VERTICAL_DISPLACEMENT = 200   // eg: Camera is 200 mm above ground
            val CAMERA_LEFT_DISPLACEMENT = 0     // eg: Camera is ON the robot's center line
            val CAMERA_CHOICE = VuforiaLocalizer.CameraDirection.BACK

            val phoneLocationOnRobot = OpenGLMatrix
                    .translation(CAMERA_FORWARD_DISPLACEMENT.toFloat(), CAMERA_LEFT_DISPLACEMENT.toFloat(), CAMERA_VERTICAL_DISPLACEMENT.toFloat())
                    .multiplied(Orientation.getRotationMatrix(
                            AxesReference.EXTRINSIC, AxesOrder.YZX, AngleUnit.DEGREES,
                            if (CAMERA_CHOICE === VuforiaLocalizer.CameraDirection.FRONT) 90.0f else -90.0f, 0.0f, 0.0f))

            for (trackable in allTrackables!!) {
                (trackable.listener as VuforiaTrackableDefaultListener).setPhoneInformation(phoneLocationOnRobot, params.cameraDirection)
            }
        }
        return allTrackables!!
    }

    override fun runOpMode() {
        telemetry.addData("Status", "Initialized")
        telemetry.update()

        val hw = Hardware(hardwareMap)

        waitForStart()
        runtime.reset()

        hw.lifter.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
        hw.lifter.mode = DcMotor.RunMode.RUN_TO_POSITION
        // goes from 0 (starting value) down to extend lifter
        hw.lifter.targetPosition = -Hardware.LIFTER_TOP_POSITION
        hw.lifter.power = 0.5
        while (hw.lifter.isBusy && opModeIsActive()) {
            idle()
        }
        hw.lifter.power = 0.0
        hw.lifter.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER // set bottom value to 0 value

        // Turn to get out of cage
        hw.rightDrive.power = -Hardware.SLOW_SPEED
        hw.leftDrive.power = Hardware.SLOW_SPEED
        sleep(1000)

        hw.rightDrive.power = 0.0
        hw.leftDrive.power = 0.0

        val detector = GoldAlignDetector()
        detector.apply {
            // Config as taken from https://github.com/MechanicalMemes/DogeCV/blob/master/Examples/GoldAlignExample.java
            init(hardwareMap.appContext, CameraViewDisplay.getInstance())
            useDefaults()
            alignSize = 100.0
            alignPosOffset = 0.0
            downscale = 0.4
            areaScoringMethod = DogeCV.AreaScoringMethod.MAX_AREA
            maxAreaScorer.weight = 0.005
            ratioScorer.weight = 5.0
            ratioScorer.perfectRatio = 1.0
            enable()
        }

        // Turn until reaching the detector
        hw.leftDrive.power = -0.5
        hw.rightDrive.power = 0.5
        while (!detector.aligned && opModeIsActive()) {
            telemetry.addLine("Gold Detector Phase")
            telemetry.addData("X pos", detector.xPosition)
            telemetry.update()

            idle()
        }
        hw.leftDrive.power = 0.5
        hw.rightDrive.power = 0.5

        sleep(100)

        hw.leftDrive.power = 0.0
        hw.rightDrive.power = 0.0

        // Initialize Vuforia tracking phase, then turn to go to depot
        val trackables = initializeVuforia()
        var lastLocation: OpenGLMatrix? = null
        while (opModeIsActive()) {
            for (trackable in trackables) {
                val listener = trackable.listener as VuforiaTrackableDefaultListener
                if (listener.isVisible) {
                    telemetry.addData("Visible target", trackable.name)
                    lastLocation = listener.updatedRobotLocation
                }
            }

            if (lastLocation != null) {
            }
        }
    }
}
